File: src/CsharpClient/QuixStreams.Streaming/States/StreamStateManager.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using Microsoft.Extensions.Logging;
using QuixStreams.State.Storage;
using QuixStreams.Streaming.Models;

namespace QuixStreams.Streaming.States
{
    /// <summary>
    /// Manages the states.
    /// </summary>
    public class StreamStateManager
    {
        private readonly ITopicConsumer topicConsumer;
        private readonly ILoggerFactory loggerFactory;
        private readonly string logPrefix;
        private readonly ILogger logger;

        private readonly object stateLock = new object();
        private readonly Dictionary<string, IStreamState> states = new Dictionary<string, IStreamState>();
        private readonly string streamId;
        
        private static readonly ConcurrentDictionary<StreamConsumerId, StreamStateManager> StreamStateManagers = new ConcurrentDictionary<StreamConsumerId, StreamStateManager>();
        
        /// <summary>
        /// The directory where the states are stored on disk. 
        /// </summary>
        public readonly string StorageDir;
        
        /// <summary>
        /// Initializes or gets an existing instance of the <see cref="StreamStateManager"/> class with the specified parameters.
        /// </summary>
        /// <param name="topicConsumer">The topic consumer used for committing state changes.</param>
        /// <param name="streamConsumerId">Stream consumer identifier information.</param>
        /// <param name="loggerFactory">The logger factory used for creating loggers.</param>
        public static StreamStateManager GetOrCreate(ITopicConsumer topicConsumer, StreamConsumerId streamConsumerId, ILoggerFactory loggerFactory)
        {
            return StreamStateManagers.GetOrAdd(streamConsumerId,
                (key) => new StreamStateManager(topicConsumer, key, loggerFactory));
        }

        /// <summary>
        /// Tries to revoke the stream state manager for the specified stream consumer id.
        /// </summary>
        /// <param name="streamConsumerId"></param>
        /// <returns></returns>
        public static bool TryRevoke(StreamConsumerId streamConsumerId)
        {
            if (StreamStateManagers.TryRemove(streamConsumerId, out var stateManager))
            {
                foreach (var stateName in stateManager.states.Keys)
                {
                    stateManager.TryDisposeStreamState(stateName);
                }

                return true;
            }

            return false;
        }
        
        private StreamStateManager(ITopicConsumer topicConsumer, StreamConsumerId streamConsumerId, ILoggerFactory loggerFactory)
        {
            this.topicConsumer = topicConsumer;
            this.loggerFactory = loggerFactory;
            this.streamId = streamConsumerId.StreamId;
            this.logPrefix = $"{streamConsumerId.ConsumerGroup}-{streamConsumerId.TopicName}-{streamConsumerId.Partition}-{streamId}";
            this.logger = this.loggerFactory.CreateLogger<StreamStateManager>();
            this.StorageDir = Path.Combine(App.GetStateStorageRootDir(), streamConsumerId.ConsumerGroup, streamConsumerId.TopicName, streamConsumerId.Partition.ToString());

            if (this.topicConsumer != null)
            {
                topicConsumer.OnDisposed += (sender, args) => { TryRevoke(streamConsumerId); };
            }
        }
        
        /// <summary>
        /// Deletes all states for the current stream.
        /// </summary>
        public void DeleteStates()
        {
            this.logger.LogTrace("Deleting Stream states for {0}", streamId);
            
            if (App.GetStateStorageType() == StateStorageTypes.RocksDb)
            {
                RocksDbStorage.DeleteStreamStates(StorageDir, streamId);
            }
            else if (App.GetStateStorageType() == StateStorageTypes.InMemory)
            {
                InMemoryStorage.DeleteStreamStates(streamId);
            }
            else
            {
                throw new ArgumentException("Invalid state storage type");
            }

            foreach (var stateName in this.states.Keys)
            {
                TryDisposeStreamState(stateName);
            }
        }

        /// <summary>
        /// Deletes the state with the specified name
        /// </summary>
        /// <param name="stateName">The state to delete</param>
        public void DeleteState(string stateName)
        {
            this.logger.LogTrace("Deleting Stream state {0} for {1}", stateName, streamId);
            TryDisposeStreamState(stateName);
            var storage = GetStateStorage(stateName);
            storage.Clear();
            storage.Dispose();
        }

        /// <summary>
        /// Creates a new application state of dictionary type with automatically managed lifecycle for the stream
        /// </summary>
        /// <param name="stateName">The name of the state</param>
        /// <returns>Dictionary stream state</returns>
        public StreamDictionaryState GetDictionaryState(string stateName)
        {
            return GetStreamState(stateName, (stateName) =>
            {
                this.logger.LogTrace("Creating Stream state for {0}", stateName);
                var storage = GetStateStorage(stateName);
                var state = new StreamDictionaryState(storage, new PrefixedLoggerFactory(this.loggerFactory, $"{logPrefix} - {stateName}"));
                return state;
            });
        }

        /// <summary>
        /// Creates a new application state of dictionary type with automatically managed lifecycle for the stream
        /// </summary>
        /// <param name="stateName">The name of the state</param>
        /// <param name="defaultValueFactory">The value factory for the state when the state has no value for the key</param>
        /// <returns>Dictionary stream state</returns>
        public StreamDictionaryState<T> GetDictionaryState<T>(string stateName,
            StreamStateDefaultValueDelegate<T> defaultValueFactory = null)
        {
            return GetStreamState(stateName, (stateName) =>
            {
                this.logger.LogTrace("Creating Stream state for {0}", stateName);
                var storage = GetStateStorage(stateName);
                var state = new StreamDictionaryState<T>(storage,
                    defaultValueFactory, new PrefixedLoggerFactory(this.loggerFactory, $"{logPrefix} - {stateName}"));
                return state;
            });
        }

        /// <summary>
        /// Creates a new application state of scalar type with automatically managed lifecycle for the stream
        /// </summary>
        /// <param name="stateName">The name of the state</param>
        /// <returns>Dictionary stream state</returns>
        public StreamScalarState GetScalarState(string stateName)
        {
            return GetStreamState(stateName, (stateName) =>
            {
                this.logger.LogTrace("Creating Stream state for {0}", stateName);
                var storage = GetStateStorage(stateName);
                var state = new StreamScalarState(storage,
                    new PrefixedLoggerFactory(this.loggerFactory, $"{logPrefix} - {stateName}"));
                return state;
            });
        }

        /// <summary>
        /// Creates a new application state of scalar type with automatically managed lifecycle for the stream
        /// </summary>
        /// <param name="stateName">The name of the state</param>
        /// <param name="defaultValueFactory">The value factory for the state when the state has no value for the key</param>
        /// <returns>Dictionary stream state</returns>
        public StreamScalarState<T> GetScalarState<T>(string stateName,
            StreamStateDefaultValueDelegate<T> defaultValueFactory = null)
        {
            return GetStreamState(stateName, (stateName) =>
            {
                this.logger.LogTrace("Creating Stream state for {0}", stateName);
                var storage = GetStateStorage(stateName);
                var state = new StreamScalarState<T>(storage,
                    defaultValueFactory, new PrefixedLoggerFactory(this.loggerFactory, $"{logPrefix} - {stateName}"));
                return state;
            });
        }

        /// <summary>
        /// Creates a new stream state of given type with automatically managed lifecycle for the stream
        /// </summary>
        /// <param name="stateName">The name of the state</param>
        /// <param name="createState">The function to create the state</param>
        /// <typeparam name="TState">The type of the stream state</typeparam>
        /// <returns>Stream state</returns>
        private TState GetStreamState<TState>(string stateName, Func<string, TState> createState)
            where TState : IStreamState
        {
            if (this.states.TryGetValue(stateName, out var existingState))
            {
                if (!(existingState is TState castedExistingState))
                {
                    throw new ArgumentException(
                        $"{logPrefix}, State '{stateName}' already exists as different stream state type.");
                }

                return castedExistingState;
            }

            lock (stateLock)
            {
                if (this.states.TryGetValue(stateName, out existingState))
                {
                    if (!(existingState is TState castedExistingState))
                    {
                        throw new ArgumentException(
                            $"{logPrefix}, State '{stateName}' already exists as different stream state type.");
                    }

                    return castedExistingState;
                }

                var state = createState(stateName);

                this.states.Add(stateName, state);
                if (this.topicConsumer == null) return state;
                var prefix = $"{logPrefix} - {stateName} | ";

                void CommittedHandler(object sender, EventArgs args)
                {
                    try
                    {
                        this.logger.LogDebug($"{prefix} | Topic committing, flushing state.");
                        state.Flush();
                        this.logger.LogTrace($"{prefix} | Topic committing, flushed state.");
                    }
                    catch (Exception ex)
                    {
                        this.logger.LogError(ex, $"{prefix} | Failed to flush state.");
                    }
                }

                this.topicConsumer.OnCommitted += CommittedHandler;

                this.topicConsumer.OnStreamsRevoked += (sender, consumers) =>
                {
                    if (consumers.All(y => y.StreamId != streamId)) return;
                    this.logger.LogDebug($"{prefix} | Stream revoked, discarding state.");
                    this.topicConsumer.OnCommitted -= CommittedHandler;
                    state.Reset();
                    TryDisposeStreamState(stateName);
                };
                return state;
            }
        }
        
        private bool TryDisposeStreamState(string stateName)
        {
            if (this.states.Remove(stateName, out var state))
            {
                state.Dispose(); // Disposes the storage
                return true;
            }

            return false;
        }
        
        private IStateStorage GetStateStorage(string stateName)
        {
            if (App.GetStateStorageType() == StateStorageTypes.RocksDb)
            {
                return RocksDbStorage.GetStateStorage(StorageDir, streamId, stateName);
            }
            else if (App.GetStateStorageType() == StateStorageTypes.InMemory)
            {
                return InMemoryStorage.GetStateStorage(streamId, stateName);
            }
            else
            {
                throw new ArgumentException("Invalid state storage type");
            }
        }
    }
}

File: src/PythonClient/src/quixstreams/states/streamstatemanager.py
import threading
import logging
from typing import Dict, Callable, Any

from .dictstreamstate import DictStreamState
from .scalarstreamstate import ScalarStreamState
from ..statestorages import IStateStorage
from ..statestorages import StreamStateType


class StreamStateManager:
    def __init__(self, stream_id: str, state_storage: IStateStorage, log_prefix: str, topic_consumer=None):
        self._topic_consumer = topic_consumer
        self._log_prefix = f"{log_prefix}{stream_id}"
        self._stream_id = stream_id
        self._state_storage = state_storage
        self._state_lock = threading.Lock()
        self._states = {}  # type: Dict[str, IStreamState]

    def get_states(self):
        return self._state_storage.get_sub_storages()

    def delete_states(self):
        # self._logger.log(logging.DEBUG, f"Deleting Stream states for {self._stream_id}")
        count = self._state_storage.delete_sub_storages()
        self._states.clear()
        return count

    def delete_state(self, state_name):
        # self._logger.log(logging.DEBUG, f"Deleting Stream state {state_name} for {self._stream_id}")
        if not self._state_storage.delete_sub_storage(state_name):
            return False
        del self._states[state_name]
        return True

    def _get_stream_state(self, state_name, create_state):
        if state_name in self._states:
            existing_state = self._states[state_name]
            if not isinstance(existing_state, create_state(state_name).__class__):
                raise Exception(
                    f"{self._log_prefix}, State '{state_name}' already exists as different stream state type.")
            return existing_state
        with self._state_lock:
            if state_name in self._states:
                existing_state = self._states[state_name]
                if not isinstance(existing_state, create_state(state_name).__class__):
                    raise Exception(
                        f"{self._log_prefix}, State '{state_name}' already exists as different stream state type.")
                return existing_state
            state = create_state(state_name)
            self._states[state_name] = state

            if self._topic_consumer is None:
                return state

            prefix = f"{self._log_prefix} - {state_name} | "

            def committed_handler(sender, args):
                try:
                    # self._logger.log(logging.DEBUG, f"{prefix} | Topic committing, flushing state.")
                    state.flush()
                    # self._logger.log(logging.DEBUG, f"{prefix} | Topic committing, flushed state.")
                except Exception as ex:
                    # self._logger.log(logging.ERROR, f"{prefix} | Failed to flush state.", ex)
                    pass

            self._topic_consumer.on_committed = committed_handler

            def on_streams_revoked(sender, consumers):
                if all(y.stream_id != self._stream_id for y in consumers):
                    return
                # self._logger.log(logging.DEBUG, f"{prefix} | Stream revoked, discarding state.")
                self._topic_consumer.on_committed = None
                state.reset()

            self._topic_consumer.on_streams_revoked = on_streams_revoked

            return state

    def get_dict_state(self, state_name: str, default_value_factory: Callable[[str], StreamStateType] = None, state_type: StreamStateType = None) -> DictStreamState[StreamStateType]:
        """
        Creates a new application state of dictionary type with automatically managed lifecycle for the stream

        Args:
            state_name: The name of the state
            state_type: The type of the state
            default_value_factory: The default value factory to create value when the key is not yet present in the state

        Example:
            >>> state_manager.get_dict_state('some_state')
            This will return a state where type is 'Any'

            >>> state_manager.get_dict_state('some_state', lambda missing_key: return {})
            this will return a state where type is a generic dictionary, with an empty dictionary as default value when
            key is not available. The lambda function will be invoked with 'get_state_type_check' key to determine type

            >>> state_manager.get_dict_state('some_state', lambda missing_key: return {}, Dict[str, float])
            this will return a state where type is a specific dictionary type, with default value

            >>> state_manager.get_dict_state('some_state', state_type=float)
            this will return a state where type is a float without default value, resulting in KeyError when not found
        """

        if state_type is None:
            state_type = Any
            # Try to figure out the type based on python typehints
            if default_value_factory is not None:
                try:
                    state_type = type(default_value_factory('get_state_type_check'))
                    if state_type is None:
                        state_type = Any
                except:
                    pass

        instance: DictStreamState = self._states.get(state_name)
        if instance is not None:
            if instance.type is not state_type:
                logging.log(logging.WARNING, f'State {state_name} already exists with a different type ({instance.type.__name__}), new type is {state_type.__name__}. Returning original state instance.')
            return instance

        instance = self._get_stream_state(state_name, lambda name: DictStreamState[StreamStateType](
            storage=self._state_storage.get_or_create_sub_storage(name), state_type=state_type, default_value_factory=default_value_factory))
        self._states[state_name] = instance
        return instance

    def get_scalar_state(self, state_name: str, default_value_factory: Callable[[], StreamStateType] = None, state_type: StreamStateType = None) -> ScalarStreamState[StreamStateType]:
        """
        Creates a new application state of scalar type with automatically managed lifecycle for the stream

        Args:
            state_name: The name of the state
            default_value_factory: The default value factory to create value when it has not been set yet
            state_type: The type of the state

        Example:
            >>> stream_consumer.get_scalar_state('some_state')
            This will return a state where type is 'Any'

            >>> stream_consumer.get_scalar_state('some_state', lambda: return 1)
            this will return a state where type is 'Any', with an integer 1 (zero) as default when
            value has not been set yet. The lambda function will be invoked with 'get_state_type_check' key to determine type

            >>> stream_consumer.get_scalar_state('some_state', lambda: return 0, float)
            this will return a state where type is a specific type, with default value

            >>> stream_consumer.get_scalar_state('some_state', state_type=float)
            this will return a state where type is a float with a default value of that type
        """

        if state_type is None:
            state_type = Any
            # Try to figure out the type based on python typehints
            if default_value_factory is not None:
                try:
                    state_type = type(default_value_factory('get_state_type_check'))
                    if state_type is None:
                        state_type = Any
                except:
                    pass

        instance: ScalarStreamState = self._states.get(state_name)
        if instance is not None:
            if instance.type is not state_type:
                logging.log(logging.WARNING, f'State {state_name} already exists with a different type ({instance.type.__name__}), new type is {state_type.__name__}. Returning original state instance.')
            return instance

        instance = self._get_stream_state(state_name, lambda name: ScalarStreamState[StreamStateType](
            storage=self._state_storage.get_or_create_sub_storage(name), state_type=state_type, default_value_factory=default_value_factory))
        self._states[state_name] = instance
        return instance

